rust-alternate/src/connection.rs

```diff
@@ -0,0 +1,256 @@
+use std::{
+    net::SocketAddr,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use anyhow::{Context as _, Result};
+use futures_util::{
+    stream::{SplitSink, SplitStream},
+    Sink, SinkExt, Stream, StreamExt,
+};
+use tokio::net::TcpStream;
+use tokio_tungstenite::{tungstenite::Message as WebsocketMessage, WebSocketStream};
+
+use crate::{
+    game_result::GameStats,
+    message::{Message, MessageType},
+};
+
+pin_project_lite::pin_project! {
+    pub struct ConnectionReadHalf {
+        addr: SocketAddr,
+        #[pin]
+        stream: SplitStream<WebSocketStream<TcpStream>>,
+    }
+}
+
+pin_project_lite::pin_project! {
+    pub struct ConnectionWriteHalf {
+        addr: SocketAddr,
+        #[pin]
+        sink: SplitSink<WebSocketStream<TcpStream>, WebsocketMessage>,
+    }
+}
+
+pin_project_lite::pin_project! {
+    pub struct Connection {
+        addr: SocketAddr,
+        #[pin]
+        read: ConnectionReadHalf,
+        #[pin]
+        write: ConnectionWriteHalf,
+    }
+}
+
+impl Connection {
+    /// Creates new connection read and write halves
+    pub fn new(addr: SocketAddr, stream: WebSocketStream<TcpStream>) -> Self {
+        let (sink, stream) = stream.split();
+
+        Self {
```

Comment:
just a side note.  I greatly dislike implicit returns.  

I have to use additional mental overhead to find see if its semi-colonless to understand its returning.  I just prefer a nice red return statement so I don't even have to consider whats happening.  General justification is that I read LTR but to know if something returns you have to read to the end of the statement then backwards pump that information.  It almost turns thing into a little RTL LTR dance.  I just find it straining (i remove all things that require an additional overhead).

Reply:
Rust coming from ML family of programming languages, and ML languages are functional

In functional language it looks like this (example is written in jsonnet.org btw):

```jsonnet
local add(a, b) = a + b;
local c = 3;
add(1, 2) + c
// and then you can treat "calls" and other things as replacements instead
local c = 3;
1 + 2 + c
// ...
local c = 3;
3 + c
//
3 + 3
//
6
```

In jsonnet (and almost all other FP languages), syntax of variable declaration is not 
`local <name> = <value>`
but 
`local <name> = <value>; <result>`

Rust inherited many of functional programming things, and those "implicit returns" shouldn't be read like so, it should be read as "result value of current expression", and "explicit returns" as "early return"

Reply:
and i still don't like implicit returns, but thank you for trying.

Reply:
But doesn't it go again your thesis in readme? 

> I want to be able to build the servers in such a way that each one is as "idiomatic"

For Rust code **it is** idiomatic to make use implicit returns whenever possible.

Moreover, it's not just about implicit returns, but about the fact, that **a lot** of things can be expressions (unlike other programming languages).

For instance, conditional expressions, `while`-loops and many others can be the one.

Not using this ability seems strange in this project: you *can* write in other code-styles, but it breaks the whole idea of using the best of languages's ecosystem and making it into just another language.

Reply:
Everybody is entitled to their own opinions, but at the end of the day we're doing this for a coding comparison run by him, that he's going to talk about on YT/Twitch.

Adding an explicit return vs an implicit one won't change the performance of the language, so I have no problem with respecting this preference. 

I don't agree with it, I like being able to return implicitly. I think it makes the code cleaner and more intuitive. But let's focus on the things that actually have an impact.

Reply:
It has no bearing on usability and there is great arguments to be made that as little processing as possible your brain needs to do leaves the smallest amount of decision fatigue.  I find it fatiguing playing discover my return.  Just a point that is pointless to argue about.

i will always bias towards as much pattern matching my brain can do.  `;` is a really hard pattern to spot in comparison to `return`.

Reply:
Well, you can't have "explicit" return here anyway, as rustfmt will replace it with "implicit" :D

Reply:
that on the other hand, very purchasable.  Don't fight the formatter